### PR TITLE
overlord, daemon: flag failures feature fancy forms.

### PR DESCRIPTION
**Pull request by chipaca**

### CODE REVIEW

File: daemon/api.go [hunk 1]
```diff
@@ -1021,37 +1021,29 @@ func (inst *snapInstruction) dispatch() snapActionFunc {
...
-	if _, ok := err.(*snap.NoUpdateAvailableError); ok {
-		return SyncResponse(&resp{
-			Type: ResponseTypeError,
-			Result: &errorResult{
-				Message: err.Error(),
-				Kind:    errorKindSnapNoUpdateAvailable,
-			},
-			Status: http.StatusBadRequest,
-		}, nil)
+	result := &errorResult{Message: err.Error()}
```
Comment (mvo5) [inline daemon/api.go]: \o/ So much nicer this way!

File: overlord/snapstate/check_snap.go [hunk 2]
```diff
@@ -99,40 +99,86 @@ func checkVersion(version string) bool {
...
-		return err
-	}
+type ErrSnapNeedsClassicSystem string
 
-	if s.NeedsDevMode() && !flags.DevModeAllowed() {
-		return fmt.Errorf("snap %q requires devmode or confinement override", s.Name())
-	}
-	if s.NeedsClassic() {
-		if !release.OnClassic {
-			return fmt.Errorf("snap %q requires classic confinement which is only available on classic systems", s.Name())
```
Comment (pedronis) [inline overlord/snapstate/check_snap.go]: where is this check gone?

File: overlord/snapstate/check_snap.go [hunk 3]
```diff
@@ -99,40 +99,86 @@ func checkVersion(version string) bool {
...
+func (e ErrSnapNeedsClassicSystem) Error() string {
+	return fmt.Sprintf("snap %q requires classic confinement which is only available on classic systems", string(e))
+}
+
+func checkSnapInfoFlags(info *snap.Info, snapst *SnapState, flags Flags) error {
```
Comment (pedronis) [inline overlord/snapstate/check_snap.go]: this probably shouldn't have a name starting with checkSnap, is really checking the flags, not the the snap

File: overlord/snapstate/check_snap.go [hunk 4]
```diff
@@ -99,40 +99,86 @@ func checkVersion(version string) bool {
...
+		return &ErrSnapNeedsMode{
+			Mode: c,
+			Snap: info.Name(),
 		}
+	default:
+		return fmt.Errorf("unknown confinement %q", c)
+	}
+}
+
+func checkSnapInfo(info *snap.Info, snapst *SnapState, flags Flags) error {
```
Comment (pedronis) [inline overlord/snapstate/check_snap.go]: this merits a comment

Comment (pedronis) [inline overlord/snapstate/check_snap.go]: also again the name makes you think it does what checkSnap does, but starting from a info, but actually it does less, so again maybe it needs a slightly different name

File: overlord/snapstate/check_snap.go [hunk 5]
```diff
@@ -99,40 +99,95 @@ func checkVersion(version string) bool {
...
-	// This assumes that the snap was already verified or --dangerous was used.
+func (e *ErrSnapNeedsMode) Error() string {
+	return fmt.Sprintf("snap %q requires %s or confinement override", e.Snap, e.Mode)
+}
 
-	s, _, err := openSnapFile(snapFilePath, si)
-	if err != nil {
-		return err
-	}
+type ErrSnapNeedsClassicSystem string
```
Comment (pedronis) [inline overlord/snapstate/check_snap.go]: this seems a bit too cute, I would have used (given that is exported) a struct with Snap anyway

**Pull request merged**